src/core/posix.cc

```diff
@@ -122,16 +122,16 @@ posix_thread::posix_thread(attr a, std::function<void ()> func)
     }
 #endif
 
-    if (a._affinity) {
-        auto& cpuset = *a._affinity;
-        pthread_attr_setaffinity_np(&pa, sizeof(cpuset), &cpuset);
-    }
-
     r = pthread_create(&_pthread, &pa,
                 &posix_thread::start_routine, _func.get());
     if (r) {
         throw std::system_error(r, std::system_category());
     }
+
+    if (a._affinity) {
+        auto& cpuset = *a._affinity;
+        pthread_setaffinity_np(_pthread, sizeof(cpuset), &cpuset);
```

Comment:
The patch looks good:
One (like me) might wonder if creating a new thread with the old affinity and only then changing the affinity doesn't have some small performance implications - like the thread being created on one CPU and then having to move. But in reality, the Linux clone() system call does **not** allow the user to control the new thread's cpu affinity, and one needs to first clone() the thread and then call sched_setaffinity. This is also what the glibc implementation does (I took a look) for pthread_attr_setaffinity_np - it just remembers it, and applies it after creating the new thread with clone(). So the fact you're doing this explicitly is perfectly fine.

Reply:
okay